pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -0,0 +1,236 @@
+		podUID:      podUID,
+	}
+}
+
+// getEvictablePods gets evictable pods.
+// Evictable pods contain sandboxes and containers which are not active and created more than MinAge ago.
+func (m *kubeGenericRuntimeManager) getEvictablePods(minAge time.Duration) (map[string]*podGCInfo, error) {
+	sandboxes, err := m.getKubeletSandboxes(true)
+	if err != nil {
+		return nil, err
+	}
+
+	evictablePods := make(map[string]*podGCInfo)
+	newestGCTime := time.Now().Add(-minAge)
+	for _, s := range sandboxes {
+		if s.GetState() != runtimeApi.PodSandBoxState_NOTREADY {
+			continue
+		}
+
+		createdAt := time.Unix(*s.CreatedAt, 0)
+		if newestGCTime.Before(createdAt) {
+			continue
+		}
+
+		podUID := getStringValueFromLabel(s.Labels, types.KubernetesPodUIDLabel)
+		if len(podUID) == 0 {
+			podUID = nonKubernetesPodUID
+		}
+		if pod, ok := evictablePods[podUID]; ok {
+			pod.sandboxIDs = append(pod.sandboxIDs, s.GetId())
+		} else {
+			evictablePods[podUID] = &podGCInfo{
+				client:     m.runtimeService,
+				containers: make([]*containerGCInfo, 0),
+				sandboxIDs: []string{s.GetId()},
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range containerStatuses {
+			evictablePods[podUID].containers = append(evictablePods[podUID].containers,
+				toContainerGCInfo(podUID, s.GetId(), c))
+		}
+
+		sort.Sort(containerByCreated(evictablePods[podUID].containers))
+	}
+
+	return evictablePods, nil
+}
+
+// enforceMaxPerPodContainer enforces max number of dead containers of each pod and
+// returns the number of total evictable containers.
+func enforceMaxPerPodContainer(evictablePods map[string]*podGCInfo, maxPerPodContainer int) int {
+	totalEvictableContainers := 0
+	for _, pod := range evictablePods {
+		if maxPerPodContainer > 0 {
+			pod.removeOldestNContainers(pod.containersNum() - maxPerPodContainer)
+		}
+		totalEvictableContainers += pod.containersNum()
+	}
+
+	return totalEvictableContainers
+}
+
+// GarbageCollect removes dead containers using the specified container gc policy.
+// It consists of following steps:
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than gcPolicy.MinAge ago.
+// * removes oldest dead containers for each pod by enforcing gcPolicy.MaxPerPodContainer.
+// * removes oldest dead containers by enforcing gcPolicy.MaxContainers.
+// * removes dead sandboxes with zero containers.
+func (m *kubeGenericRuntimeManager) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy, allSourcesReady bool) error {
+	evictablePods, err := m.getEvictablePods(gcPolicy.MinAge)
+	if err != nil {
+		glog.Warningf("getEvictablePods failed: %v", err)
+		return err
+	}
+
+	glog.V(4).Infof("GarbageCollect gets pods %q for evicting", evictablePods)
+
+	// Enforce max number of dead containers of each pod
+	totalEvictableContainers := enforceMaxPerPodContainer(evictablePods, gcPolicy.MaxPerPodContainer)
```

Comment:
ah..I see you are just want to return the rest of the containers that are evictable. It's not very clear from just reading the description of the function.